Fix input sanitation for listchaintxns lncli cmd

kornpow opened the pull request:
## Change Description
This PR was based originally on the PR #7496 and issue #7316 .

Upon rebasing and looking at the solutions there, I found the bigger problem with the command was the lack of error messaging when the `end_height < start_height` rather than the necessity to support a "reverse ordering" feature.

So while in a separate soon so be opened PR I also implemented "reverse ordering" I thought I would open this one first, as it is more straightforward and an improvement necessary, whether or not reverse ordering is needed/wanted.

This is my first code commit to LND, 2nd including a comment update, so I appreciate the help to getting this PR mergable!

## Steps to Test
Try using different options for `start_height` and `end_height`, including them, not including them, and having integer values in a way which doesn't make sense, and ensure that the proper errors are shown.


## Pull Request Checklist
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ] Any new lncli commands have appropriate tags in the comments for the rpc in the proto file.
- [ ] [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

guggero commented on cmd/commands/commands.go:
```diff
@@ -2267,10 +2267,24 @@ func listChainTxns(ctx *cli.Context) error {
+
+		if endHeight != -1 && startHeight > endHeight {
+			return errors.New("start_height should " +
+				"be less than end_height if end_height" +
```
nit: always put space between words on previous line for consistency. If the line becomes too long, break a word earlier.

guggero commented on docs/release-notes/release-notes-0.19.0.md:
```diff
@@ -211,8 +211,13 @@
 `ProofMatureDelta` used in gossip to be configurable via
 `--gossip.announcement-conf`, with a default value of 6.
 
+
```
nit: unnecessary newlines?

guggero commented on docs/release-notes/release-notes-0.19.0.md:
```diff
@@ -211,8 +211,13 @@
+
 ## lncli Updates
 
+* [Fix](https://github.com/lightningnetwork/lnd/pull/9558) Added additional 
```
nit: Probably want to replace "Fix" with "Added" so it's a complete sentence?

kornpow replied:
Thanks, I see the pattern now :)

guggero reviewed (commented):
Thanks for the fix. Makes sense, I think the "will be queried in reverse" was probably from an earlier version of the PR that added the offset.

kornpow commented:
@guggero I would like to potentially add a test for this, but I am not really sure where? Maybe there isnt tests for these kinds of things.

lightninglabs-deploy commented:
@guggero: review reminder

guggero commented:
!lightninglabs-deploy mute

kornpow commented:
I have added a test with a decent number of test cases.

I also modified the `listchaintxns` command to formalize that the default case value for `end_height=-1`. This part confused me a bit when reading through the code, since things "seem to work" when `end_height` is left unset, which used to give it a default value of 0. However, I think setting it to a default is better.

Finally, as much as I would like to remove the cli context from the test itself, since it makes the test code a decent bit more verbose, I think it is necessary because of testing the behavior like in the paragraph above, where there are certain values unset in the CLI. Also I think it is good because it could provide an example for others how to test the CLI.

Please let me know if you would like me to improve the test structure. I potentially could make the inputs inside a loop and reusing variables, but there is funkiness there with redefining values in flag which I'd have to tackle.

kornpow commented:
Been running my new test like this:
`env GOEXPERIMENT=loopvar go test -tags="dev autopilotrpc chainrpc invoicesrpc neutrinorpc peersrpc routerrpc signrpc verrpc walletrpc watchtowerrpc wtclientrpc nolog" github.com/lightningnetwork/lnd/cmd/commands -v -run "TestInputListChainTxns"`

MPins commented on cmd/commands/commands_test.go:
```diff
@@ -328,3 +330,145 @@ func TestAppendChanID(t *testing.T) {
+	var startHeight, endHeight int64
+	flagSet1.Int64Var(&startHeight, "start_height", startDefault, "")
+	flagSet1.Int64Var(&endHeight, "end_height", endDefault, "")
+	flagSet1.Set("start_height", "100")
```
This function has a return value, and you need to check it to pass the lint check.

Same to others call below.

MPins reviewed (commented):
Good job for your first PR :+1: 

I ran the test and it’s working — maybe just a few small adjustments are needed.

kornpow commented on cmd/commands/commands_test.go:
```diff
@@ -328,3 +330,145 @@ func TestAppendChanID(t *testing.T) {
+	flagSet4.Int64Var(&endHeight4, "end_height", endDefault, "")
+	flagSet4.Set("start_height", "100")
+	flagSet4.Set("end_height", "-1")
+	ctx4 := cli.NewContext(app, flagSet4, nil)
+
+	// Case 5: start set and end height set as 0
+	flagSet5 := flag.NewFlagSet("listchaintxns", flag.ContinueOnError)
+	var startHeight5, endHeight5 int64
+	flagSet5.Int64Var(&startHeight5, "start_height", startDefault, "")
+	flagSet5.Int64Var(&endHeight5, "end_height", endDefault, "")
+	flagSet5.Set("start_height", "100")
+	flagSet5.Set("end_height", "0")
```
It is pretty much the same I agree...

I might have written this before I set the default value of `end_height=-1`. I can probably remove this case.

kornpow commented:
I also made a commit with some changes to the help text of the command. It feels like too much of the `Flag Usage` is finding its way into the `Description`. Its also really redundant trying to get across the `end_height=-1` concept. Let me know if that is out of scope.

MPins commented on cmd/commands/commands_test.go:
```diff
@@ -328,3 +330,100 @@ func TestAppendChanID(t *testing.T) {
 		})
 	}
 }
+
```
nit: doc the function (see the other functions above)

kornpow replied:
Added ty

MPins reviewed (commented):
Great progress! :+1: 

I’d suggest squashing the commits into two: one for the doc release and one for the other files — what do you think?

https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure

kornpow commented:
> Great progress! 👍
> 
> I’d suggest squashing the commits into two: one for the doc release and one for the other files — what do you think?
> 
> https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure

Sounds good. Squashed to 2 commits. Getting better at managing the commit history, but still learning where to split the commits. Getting closer I hope!

MPins commented:
> > Great progress! 👍
> > I’d suggest squashing the commits into two: one for the doc release and one for the other files — what do you think?
> > https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#ideal-git-commit-structure
> 
> Sounds good. Squashed to 2 commits. Getting better at managing the commit history, but still learning where to split the commits. Getting closer I hope!

It would be a good idea to have some description on commit message. Something like:

```
commands: add a parsing block height function
Added parseBlockHeightInputs function to parse start_height and end_height
from the CLI, ensuring start_height is non-negative and not greater than end_height
(unless it's -1), returning both values and any validation error.
```

MPins commented on cmd/commands/commands_test.go:
```diff
@@ -328,3 +330,102 @@ func TestAppendChanID(t *testing.T) {
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			flagSet1 := flag.NewFlagSet("listchaintxns",
```
nit: I think `flagSet` is better than `flagSet1`

MPins reviewed (approved):
Some final comments and LGTM! :+1: 

It would be a good idea to have some description on commit message. Something like:

```
commands: add a parsing block height function
Added parseBlockHeightInputs function to parse start_height and end_height
from the CLI, ensuring start_height is non-negative and not greater than end_height
(unless it's -1), returning both values and any validation error.
```

kornpow commented:
rebased to fix a merge conflict and to improve commit message.

MPins commented:
> rebased to fix a merge conflict and to improve commit message.

You should leave just one of the following lines in the commit message:
```
cmd: Modified listchaintxns command

commands: add a parsing block height function
```
Maybe the first is better :wink:

kornpow commented:
> > rebased to fix a merge conflict and to improve commit message.
> 
> You should leave just one of the following lines in the commit message:
> 
> ```
> cmd: Modified listchaintxns command
> 
> commands: add a parsing block height function
> ```
> 
> Maybe the first is better 😉

Rebased again onto master, and modified the commit message... Hopefully acceptable now.

Its been hard to get all of the integration tests to always pass consistently? I don't think it is related to my changes though? Please let me know.

guggero commented on cmd/commands/commands_test.go:
```diff
@@ -328,3 +330,102 @@ func TestAppendChanID(t *testing.T) {
+
+			err := flagSet.Set("start_height",
+				strconv.Itoa(int(tc.expectedStart)))
+			if err != nil {
```
We should use `require.NoError()` for new test code.

guggero reviewed (commented):
Thanks for the updates. Getting closer.

kornpow commented on cmd/commands/commands.go:
```diff
@@ -2224,24 +2224,27 @@ var listChainTxnsCommand = cli.Command{
+				"transactions, inclusive; to get " +
 				"transactions until the chain tip, including " +
 				"unconfirmed, set this value to -1",
+			Value: -1,
```
What is incredibly confusing is that currently setting it at 0 -> ie not setting it at all -> still returns unconfirmed transactions.

I find it much more confusing that both 0 and -1 returns unconfirmed, which is why I made the change.

kornpow replied:
This is where that happens:
https://github.com/lightningnetwork/lnd/blob/master/rpcserver.go#L6563-L6568
```
	var endHeight = btcwallet.UnconfirmedHeight

	// If the user has provided an end height, we overwrite our default.
	if req.EndHeight != 0 {
		endHeight = req.EndHeight
	}
```
This code says if we leave the default as 0, then 
`endHeight = btcwallet.UnconfirmedHeight`

guggero replied:
Ah, okay... Yeah, default values are notoriously hard to get right in the command line. So I guess makes sense to set `-1` here but perhaps add a comment above it that mentions something along the lines of "Will be set to this value by the server if left blank".

kornpow replied:
Ok pushing up a change to the flag usage text which I think makes it clearer.

Ideally someday we can get rid of this `-1` (C-like) special case behavior and just add a boolean flag to enable unconfirmed txns. The boolean can even be default true so the behavior is pretty much the same.

guggero replied:
> The boolean can even be default true 

No with the current CLI library we use :joy:  You can't set a boolean flag to `false`. So yeah, things are tricky. And moving to a different CLI library is a huuuuge lift...

kornpow commented:
![image](https://github.com/user-attachments/assets/1d3247b4-0785-45ba-88f2-be057db6d018)
Picture of the new help text, to get a feel for how it looks in-situ
